client/rest-high-level/src/test/java/org/elasticsearch/client/CustomRestHighLevelClientTests.java

```diff
@@ -0,0 +1,236 @@
+    }
+
+    private Void performRequestAsync(HttpEntity httpEntity, ResponseListener responseListener) {
+        try {
+            responseListener.onSuccess(performRequest(httpEntity));
+        } catch (IOException e) {
+            responseListener.onFailure(e);
+        }
+        return null;
+    }
+
+    private Response performRequest(HttpEntity httpEntity) throws IOException {
+        try (XContentParser parser = createParser(REQUEST_BODY_CONTENT_TYPE.xContent(), httpEntity.getContent())) {
+            CustomRequest request = CustomRequest.fromXContent(parser);
+
+            int length = request.getValue() != null ? request.getValue().length() : -1;
+            CustomResponse response = new CustomResponse(length);
+
+            ProtocolVersion protocol = new ProtocolVersion("HTTP", 1, 1);
+            RestStatus status = response.status();
+            HttpResponse httpResponse = new BasicHttpResponse(new BasicStatusLine(protocol, status.getStatus(), status.name()));
+
+            BytesRef bytesRef = XContentHelper.toXContent(response, XContentType.JSON, false).toBytesRef();
+            httpResponse.setEntity(new ByteArrayEntity(bytesRef.bytes, ContentType.APPLICATION_JSON));
+
+            RequestLine requestLine = new BasicRequestLine(HttpGet.METHOD_NAME, ENDPOINT, protocol);
+            return new Response(requestLine, new HttpHost("localhost", 9200), httpResponse);
+        }
+    }
+
+    /**
+     * A custom high level client that provides methods to execute a custom request and get its associate response back.
+     */
+    static class CustomRestClient extends RestHighLevelClient {
+
+        private CustomRestClient(RestClient restClient) {
+            super(restClient);
+        }
+
+        public CustomResponse custom(CustomRequest customRequest, Header... headers) throws IOException {
+            return performRequest(customRequest, this::toRequest, this::toResponse, emptySet(), headers);
+        }
+
+        public void customAsync(CustomRequest customRequest, ActionListener<CustomResponse> listener, Header... headers) {
+            performRequestAsync(customRequest, this::toRequest, this::toResponse, listener, emptySet(), headers);
+        }
+
+        Request toRequest(CustomRequest customRequest) throws IOException {
+            BytesRef source = XContentHelper.toXContent(customRequest, REQUEST_BODY_CONTENT_TYPE, false).toBytesRef();
+            ContentType contentType = ContentType.create(REQUEST_BODY_CONTENT_TYPE.mediaType());
+            HttpEntity entity = new ByteArrayEntity(source.bytes, source.offset, source.length, contentType);
+            return new Request(HttpGet.METHOD_NAME, ENDPOINT, emptyMap(), entity);
+        }
+
+        CustomResponse toResponse(Response response) throws IOException {
+            return parseEntity(response.getEntity(), CustomResponse::fromXContent);
+        }
+    }
+
+    static class CustomRequest extends ActionRequest implements ToXContentObject {
```

Comment:
I'd like to keep the custom request and custom response because it echoes the custom/customAsync methods. But I can trim down the test by not printing out xcontent request and response body which are not the point here. I'll try that and you'll tell me if you still want to reuse existing requests (I'm fine with reusing MainRequest/MainResponse for example)